api/python/quilt3/packages.py

```diff
@@ -1338,3 +1338,37 @@ def _filter(self, f, include_directories=False):
                 p._set(lk, entity)
 
         return p
+
+    def verify(self, src, extra_files=False):
+        """
+        Check if the contents of the given directory matches the package manifest.
+
+        Args:
+            src(str): URL of the directory
+            extra_files(bool): Whether extra files in the directory should cause a failure.
+        Returns:
+            True if the package matches the directory; False otherwise.
+        """
+        src = fix_url(src).rstrip('/') + '/'
+        src_dict = dict(list_url(src))
+        url_list = []
+        size_list = []
+        for logical_key, entry in self.walk():
+            src_size = src_dict.pop(logical_key, None)
+            if src_size is None:
+                return False
+            if entry.size != src_size:
+                return False
+            entry_url = src + quote(logical_key)
+            url_list.append(entry_url)
+            size_list.append(src_size)
+
+        if src_dict and not extra_files:
+            return False
+
+        hash_list = calculate_sha256(url_list, size_list)
```

Comment:
Ideally we'd have "short-circuit" version of `calculate_sha256` but that can be done in a separate PR. i.e. this will do all hashes even if all but one of them are wrong. Expected-case performance would be much better with short circuit.